xen/arch/x86/smpboot.c

```diff
@@ -322,41 +315,24 @@ static void set_cpu_sibling_map(unsigned int cpu)
     }
 }
 
-void start_secondary(void *unused)
+void start_secondary(unsigned int cpu)
 {
     struct cpu_info *info = get_cpu_info();
 
+    /* Tell BSP that we are awake. */
+    set_cpu_state(cpu, CPU_STATE_INIT);
+
     /*
-     * Dont put anything before smp_callin(), SMP booting is so fragile that we
+     * Don't put anything before smp_callin(), SMP booting is so fragile that we
      * want to limit the things done here to the most necessary things.
      */
-    unsigned int cpu = booting_cpu;
-
-    if ( sl_status ) {
-        uint64_t misc_enable;
-        uint32_t my_apicid;
-        struct txt_sinit_mle_data *sinit_mle =
-              txt_sinit_mle_data_start(__va(read_txt_reg(TXTCR_HEAP_BASE)));
-
-        /* TXT released us with MONITOR disabled in IA32_MISC_ENABLE. */
-        rdmsrl(MSR_IA32_MISC_ENABLE, misc_enable);
-        wrmsrl(MSR_IA32_MISC_ENABLE,
-               misc_enable | MSR_IA32_MISC_ENABLE_MONITOR_ENABLE);
-
-        /* get_apic_id() reads from x2APIC if it thinks it is enabled. */
-        x2apic_ap_setup();
-        my_apicid = get_apic_id();
-
-        while ( my_apicid != x86_cpu_to_apicid[cpu] ) {
-            asm volatile ("monitor; xor %0,%0; mwait"
-                          :: "a"(__va(sinit_mle->rlp_wakeup_addr)), "c"(0),
-                          "d"(0) : "memory");
```

Comment:
No longer needed? Why?

Reply:
This was done to hold APs here until all notifiers are called. Now they just loop on `while ( cpu_data[cpu].cpu_state != CPU_STATE_CALLOUT )`, both for TXT and non-TXT path.